README.md
#node-red-contrib-gps-lite-gmiot

万物在线节点


## 安装

**Requires Node.js v6.0 or newer**

Either install from the Node-RED palette manager, or:

```
$ npm i node-red-contrib-gps-lite-gmiot
```

## 使用

* **gmiot-login** 用于登陆万物在线，目前`access_token`有效期为2小时，所以要定时获取 
* **gmiot-monitor** 用于获取设备信息，默认`msg.payload.data`是第一个设备信息，详细设备可以从`msg.payload.dataArray`获取
* **gmiot-address** 通过经纬度反向获取地址信息，输入参数经纬度


## QQ群 776817275 欢迎大家加入

![](https://i.loli.net/2018/12/28/5c25b8bf1e78d.jpg)

## Maintainers

[@yaming116](https://github.com/yaming116)

## License

MIT © 2018 yaming116



changelog.md
* 1.0.0
满足基本使用


login.js
module.exports = function (RED) {
    var axios = require('axios');
    var CryptoJS = require("crypto-js");

    function login(config) {
        RED.nodes.createNode(this, config);

        // Retrieve the config node
        this.server = RED.nodes.getNode(config.server);
        
        var node = this;
        if (this.server) {
            node.status({fill:"blue", shape:"ring", text: "ok"});
        } else {
            node.status({fill:"red", shape:"ring", text: "没有配置正确的万物在线信息"});
            return
        }
        let phone = this.server.phone
        this.sid = phone;
        let md5Pass = this.server.md5Pass

        node.on('input', function (msg) {
            var time = parseInt(new Date().getTime() / 1000, 10)

            var payload = {}
            
            var signature = CryptoJS.MD5(md5Pass + time).toString()

            console.log(`account: ${phone} time: ${time} md5Pass: ${md5Pass}`)
            console.log(`http://litapi.gmiot.net/1/auth/access_token?account=${phone}&time=${time}&signature=${signature}`)
            axios({
                method: 'get',
                url: `http://litapi.gmiot.net/1/auth/access_token?account=${phone}&time=${time}&signature=${signature}`,
                headers: {
                    'Content-type': 'Content-type: text/html; charset=utf-8'
                }


            }).then(function (response) {
                var data = response.data
                if(data.ret && data.ret != 0) {
                    throw new Error(data.msg ? data.msg : "参数有误")
                }
                msg.headers = response.headers
                msg.request = response.request
                payload.status = 1
                payload.data = data
                let global = node.context().global
                global.set(phone, {access_token: data.access_token, expires_in: data.expires_in, time: parseInt(new Date().getTime() / 1000, 10)})
                
                msg.payload = payload
                console.log(11111111111111)
                node.send([msg, null])

            }).catch(function (error) {
                console.log(222222222222222222)
                payload.status = -1
                payload.data = error.message
                msg.payload = payload
                msg['error'] = error
                node.send([null, msg])
            })
        });

    }
    RED.nodes.registerType("gmiot-login", login);

}

monitor.js
module.exports = function (RED) {
    var axios = require('axios');

    function monitor(config) {
        RED.nodes.createNode(this, config);

        // Retrieve the config node
        this.server = RED.nodes.getNode(config.server);
        
        var node = this;
        if (this.server) {
            node.status({fill:"blue", shape:"ring", text: "ok"});
        } else {
            node.status({fill:"red", shape:"ring", text: "没有配置正确的万物在线信息"});
            return
        }

        node.on('input', function (msg) {
            getDataFromHttp(node, node.server, msg)
        });
    }

    function unlogin(node , msg) {
        var payload = {}
        payload.status = 10006
        payload.data = 'access_token 信息不存在，请重新登陆'
        msg.payload = payload
        node.send([null, msg])
    }

    function getDataFromHttp(node , server, msg){
        var payload = {}
        var global = node.context().global
        if (!global.get(server.phone)) {
            unlogin(node, msg)
            return
        }
        var access_token = global.get(server.phone).access_token
        console.log(`target: ${server.phone} access_token： ${access_token}`)
        node.status({fill:"yellow", shape:"ring", text: "request..."});

        axios({
            method: 'get',
            url: `http://litapi.gmiot.net/1/account/monitor?target=${server.phone}&access_token=${access_token}`,
            headers: {
                'Content-type': 'Content-type: text/html; charset=utf-8',
            }

        }).then(function (response) {
            node.status({fill:"blue", shape:"ring", text: "ok"});

            var data = response.data
            console.log(JSON.stringify(data))
            if(data.ret > 10003 && data.ret < 10007) {
                unlogin(node, msg)
                return
            }
            if(data.ret && data.ret != 0) {
                throw new Error(data)
            }

            msg.headers = response.headers
            msg.request = response.request
            payload.status = 1
            payload.dataArray = data

            if(data.data.length > 0) {
                payload.data = data.data[0]
            }

            msg.payload = payload
            node.send([msg, null])

        }).catch(function (error) {
            node.status({fill:"blue", shape:"ring", text: "error"});

            payload.status = -1
            payload.data = error.message
            msg.payload = payload
            msg['error'] = error
            node.send([null, msg])
        })
    }
    RED.nodes.registerType("gmiot-monitor", monitor);

}

address.js
module.exports = function (RED) {
    var axios = require('axios');

    function address(config) {
        RED.nodes.createNode(this, config);

        // Retrieve the config node
        this.server = RED.nodes.getNode(config.server);
        
        var node = this;
        if (this.server) {
            node.status({fill:"blue", shape:"ring", text: "ok"});
        } else {
            node.status({fill:"red", shape:"ring", text: "没有配置正确的万物在线信息"});
            return
        }

        node.on('input', function (msg) {
            getDataFromHttp(node, node.server, msg, config)
        });
    }

    function unlogin(node , msg) {
        var payload = {}
        payload.status = 10006
        payload.data = 'access_token 信息不存在，请重新登陆'
        msg.payload = payload
        node.send([null, msg])
    }

    function getDataFromHttp(node , server, msg, config){
        var longitude = config.longitude || msg.payload.longitude
        var latitude = config.latitude || msg.payload.latitude

        var payload = {}
        var global = node.context().global
        if (!global.get(server.phone)) {
            unlogin(node, msg)
            return
        }
        var access_token = global.get(server.phone).access_token
        console.log(`target: ${server.phone} access_token： ${access_token}`)
        node.status({fill:"yellow", shape:"ring", text: "request..."});
        axios({
            method: 'get',
            url: `http://litapi.gmiot.net/1/tool/address?lng=${longitude}&lat=${latitude}&access_token=${access_token}`,
            headers: {
                'Content-type': 'Content-type: text/html; charset=utf-8',
            }

        }).then(function (response) {
            node.status({fill:"blue", shape:"ring", text: "ok"});

            var data = response.data
            console.log(JSON.stringify(data))
            if(data.ret > 10003 && data.ret < 10007) {
                unlogin(node, msg)
                return
            }
            if(data.ret && data.ret != 0) {
                throw new Error(data)
            }

            msg.headers = response.headers
            msg.request = response.request
            payload.status = 1
            payload.data = data

            msg.payload = payload
            node.send([msg, null])

        }).catch(function (error) {
            node.status({fill:"blue", shape:"ring", text: "error"});

            payload.status = -1
            payload.data = error.message
            msg.payload = payload
            msg['error'] = error
            node.send([null, msg])
        })
    }
    RED.nodes.registerType("gmiot-address", address);

}